Tests with explicit nulls are broken on main branch

It seems the CI is failing on main branch because of the PR #22473.

The errors only happen in "Test Suite with the CC Standard Library". Maybe related to cc library compiled in explicit nulls?

One of the failed tests is a debug test introduced by #22597, but the error message does not give much information...

```scala
Error:  Debug step failed: tests/debug-custom-args/eval-explicit-nulls.check:3
Evaluation failed:
-- [E170] Type Error: <expression class>:12:28 -------------------------------------------------------------------------
12 |    if idx == -1 then throw new NoSuchElementException(name)
   |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                            NoSuchElementException is not a class type
   |
   | longer explanation available when compiling with `-explain`-- [E170] Type Error: <expression class>:18:28 -------------------------------------------------------------------------
18 |    if idx == -1 then throw new NoSuchElementException(name)
   |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                            NoSuchElementException is not a class type
   |
   | longer explanation available when compiling with `-explain`
Error:  Test dotty.tools.debug.DebugTests.debug failed: java.lang.AssertionError: Debug test failed, but should not, reasons:
Error:  encountered 1 test failure(s):
Error:    - generic failure (see test output), took 74.963 sec
Error:      at dotty.tools.vulpix.ParallelTesting$CompilationTest.checkPass(ParallelTesting.scala:1269)
Error:      at dotty.tools.debug.DebugTests$.dotty$tools$debug$DebugTests$$$checkDebug(DebugTests.scala:45)
Error:      at dotty.tools.debug.DebugTests.debug(DebugTests.scala:23)
Error:      at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
Error:      at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
Error:      at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
Error:      at java.lang.reflect.Method.invoke(Method.java:569)
Error:      ...
```

cc @HarrisL2 @olhotak

--- a/compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala
+++ b/compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala
@@ -55,7 +55,7 @@ object JavaNullInterop {
    */
   def nullifyMember(sym: Symbol, tp: Type, isEnumValueDef: Boolean)(using Context): Type = trace(i"nullifyMember ${sym}, ${tp}"){
     assert(ctx.explicitNulls)
-    // assert(sym.is(JavaDefined), "can only nullify java-defined members")
+    assert(sym.is(JavaDefined), "can only nullify java-defined members")
 
     // Some special cases when nullifying the type
     if isEnumValueDef || sym.name == nme.TYPE_ then

--- a/compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala
+++ b/compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala
@@ -35,7 +35,7 @@ import dotty.tools.dotc.core.Decorators.i
  *   to handle the full spectrum of Scala types. Additionally, some kinds of symbols like constructors and
  *   enum instances get special treatment.
  */
-object ImplicitNullInterop {
+object JavaNullInterop {
 
   /** Transforms the type `tp` of Java member `sym` to be explicitly nullable.
    *  `tp` is needed because the type inside `sym` might not be set when this method is called.
@@ -55,6 +55,7 @@ object ImplicitNullInterop {
    */
   def nullifyMember(sym: Symbol, tp: Type, isEnumValueDef: Boolean)(using Context): Type = trace(i"nullifyMember ${sym}, ${tp}"){
     assert(ctx.explicitNulls)
+    // assert(sym.is(JavaDefined), "can only nullify java-defined members")
 
     // Some special cases when nullifying the type
     if isEnumValueDef || sym.name == nme.TYPE_ // Don't nullify the `TYPE` field in every class and Java enum instances
@@ -80,14 +81,14 @@ object ImplicitNullInterop {
    *  but the result type is not nullable.
    */
   private def nullifyExceptReturnType(tp: Type)(using Context): Type =
-    new ImplicitNullMap(outermostLevelAlreadyNullable = true)(tp)
+    new JavaNullMap(outermostLevelAlreadyNullable = true)(tp)
 
-  /** Nullifies a type by adding `| Null` in the relevant places. */
+  /** Nullifies a Java type by adding `| Null` in the relevant places. */
   private def nullifyType(tp: Type)(using Context): Type =
-    new ImplicitNullMap(outermostLevelAlreadyNullable = false)(tp)
+    new JavaNullMap(outermostLevelAlreadyNullable = false)(tp)
 
-  /** A type map that implements the nullification function on types. Given a Java-sourced type or an
-   *  implicitly null type, this adds `| Null` in the right places to make the nulls explicit.
+  /** A type map that implements the nullification function on types. Given a Java-sourced type, this adds `| Null`
+   *  in the right places to make the nulls explicit in Scala.
    *
    *  @param outermostLevelAlreadyNullable whether this type is already nullable at the outermost level.
    *                                       For example, `Array[String] | Null` is already nullable at the
@@ -97,7 +98,7 @@ object ImplicitNullInterop {
    *                                       This is useful for e.g. constructors, and also so that `A & B` is nullified
    *                                       to `(A & B) | Null`, instead of `(A | Null & B | Null) | Null`.
    */
-  private class ImplicitNullMap(var outermostLevelAlreadyNullable: Boolean)(using Context) extends TypeMap {
+  private class JavaNullMap(var outermostLevelAlreadyNullable: Boolean)(using Context) extends TypeMap {
     def nullify(tp: Type): Type = if ctx.flexibleTypes then FlexibleType(tp) else OrNull(tp)
 
     /** Should we nullify `tp` at the outermost level? */

--- a/compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala
+++ b/compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala
@@ -58,9 +58,8 @@ object JavaNullInterop {
     // assert(sym.is(JavaDefined), "can only nullify java-defined members")
 
     // Some special cases when nullifying the type
-    if isEnumValueDef || sym.name == nme.TYPE_ // Don't nullify the `TYPE` field in every class and Java enum instances
-    || sym.is(Flags.ModuleVal) // Don't nullify Modules
-    then
+    if isEnumValueDef || sym.name == nme.TYPE_ then
+      // Don't nullify the `TYPE` field in every class and Java enum instances
       tp
     else if sym.name == nme.toString_ || sym.isConstructor || hasNotNullAnnot(sym) then
       // Don't nullify the return type of the `toString` method.

--- a/compiler/src/dotty/tools/dotc/core/Types.scala
+++ b/compiler/src/dotty/tools/dotc/core/Types.scala
@@ -3432,7 +3432,6 @@ object Types extends TypeUtils {
         // flexible type is always a subtype of the original type and the Object type.
         // It is not necessary according to the use cases, so we choose to use a simpler
         // rule.
-        assert(!tp.isInstanceOf[LazyType])
         FlexibleType(OrNull(tp), tp)
     }
   }

--- a/compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala
+++ b/compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala
@@ -525,7 +525,7 @@ class ClassfileParser(
       denot.info = translateTempPoly(attrCompleter.complete(denot.info, isVarargs))
       if (isConstructor) normalizeConstructorInfo()
 
-      if (ctx.explicitNulls) denot.info = ImplicitNullInterop.nullifyMember(denot.symbol, denot.info, isEnum)
+      if (ctx.explicitNulls) denot.info = JavaNullInterop.nullifyMember(denot.symbol, denot.info, isEnum)
 
       // seal java enums
       if (isEnum) {

--- a/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
+++ b/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
@@ -981,12 +981,6 @@ class TreeUnpickler(reader: TastyReader,
           sym.info = tpt.tpe
           ValDef(tpt)
       }
-
-      // If explicit nulls is enabled, and the source file did not have explicit
-      // nulls enabled, nullify the member to allow for compatibility.
-      if (ctx.explicitNulls && !explicitNulls) then
-        sym.info = JavaNullInterop.nullifyMember(sym, sym.info, sym.is(Enum))
-
       goto(end)
       setSpan(start, tree)
 

--- a/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
+++ b/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
@@ -985,7 +985,7 @@ class TreeUnpickler(reader: TastyReader,
       // If explicit nulls is enabled, and the source file did not have explicit
       // nulls enabled, nullify the member to allow for compatibility.
       if (ctx.explicitNulls && !explicitNulls) then
-        sym.info = ImplicitNullInterop.nullifyMember(sym, sym.info, sym.is(Enum))
+        sym.info = JavaNullInterop.nullifyMember(sym, sym.info, sym.is(Enum))
 
       goto(end)
       setSpan(start, tree)

--- a/compiler/src/dotty/tools/dotc/typer/Namer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Namer.scala
@@ -1902,7 +1902,7 @@ class Namer { typer: Typer =>
 
     val mbrTpe = paramFn(checkSimpleKinded(typedAheadType(mdef.tpt, tptProto)).tpe)
     if (ctx.explicitNulls && mdef.mods.is(JavaDefined))
-      ImplicitNullInterop.nullifyMember(sym, mbrTpe, mdef.mods.isAllOf(JavaEnumValue))
+      JavaNullInterop.nullifyMember(sym, mbrTpe, mdef.mods.isAllOf(JavaEnumValue))
     else mbrTpe
   }
 

--- a/compiler/test/dotty/tools/dotc/CompilationTests.scala
+++ b/compiler/test/dotty/tools/dotc/CompilationTests.scala
@@ -217,18 +217,8 @@ class CompilationTests {
       compileFilesInDir("tests/explicit-nulls/pos", explicitNullsOptions),
       compileFilesInDir("tests/explicit-nulls/flexible-types-common", explicitNullsOptions),
       compileFilesInDir("tests/explicit-nulls/unsafe-common", explicitNullsOptions and "-language:unsafeNulls" and "-Yno-flexible-types"),
-    ).checkCompile()
-
-    locally {
-      val tests = List(
-        compileFile("tests/explicit-nulls/flexible-unpickle/Unsafe_1.scala", explicitNullsOptions without "-Yexplicit-nulls"),
-        compileFile("tests/explicit-nulls/flexible-unpickle/Flexible_2.scala", explicitNullsOptions.withClasspath(
-          defaultOutputDir + testGroup + "/Unsafe_1/flexible-unpickle/Unsafe_1")),
-      ).map(_.keepOutput.checkCompile())
-
-      tests.foreach(_.delete())
-    }
-  }
+    )
+  }.checkCompile()
 
   @Test def explicitNullsWarn: Unit = {
     implicit val testGroup: TestGroup = TestGroup("explicitNullsWarn")

--- a/tests/explicit-nulls/flexible-unpickle/Flexible_2.scala
+++ b/tests/explicit-nulls/flexible-unpickle/Flexible_2.scala
@@ -1,7 +1,0 @@
-@main
-def Flexible_2() =
-  val s2: String | Null = "foo"
-  val unsafe = new Unsafe_1()
-  val s: String = unsafe.foo(s2)
-  unsafe.foo("")
-  unsafe.foo(null)

--- a/tests/explicit-nulls/flexible-unpickle/Flexible_2.scala
+++ b/tests/explicit-nulls/flexible-unpickle/Flexible_2.scala
@@ -1,6 +1,3 @@
-import unsafeNulls.Foo.*
-import unsafeNulls.Unsafe_1
-
 @main
 def Flexible_2() =
   val s2: String | Null = "foo"

--- a/tests/explicit-nulls/flexible-unpickle/Unsafe_1.scala
+++ b/tests/explicit-nulls/flexible-unpickle/Unsafe_1.scala
@@ -1,12 +1,6 @@
-package unsafeNulls
-
 class Unsafe_1 {
   def foo(s: String): String = {
     if (s == null) then "nullString"
     else s
   }
-}
-
-object Foo {
-  def bar = "bar!"
 }

--- a/tests/explicit-nulls/flexible-unpickle/Unsafe_1.scala
+++ b/tests/explicit-nulls/flexible-unpickle/Unsafe_1.scala
@@ -1,6 +1,0 @@
-class Unsafe_1 {
-  def foo(s: String): String = {
-    if (s == null) then "nullString"
-    else s
-  }
-}

